File: J9Ch/src/J_7_Two_Pointers/Required_10/_533_Two_Sum_Closest_to_target_Medium.py
'''/** 533 Two Sum - Closest to target
 * Medium'''


class Solution:
    # @param {int[]} nums an integer array
    # @param {int} target an integer
    # @return {int} the difference between the sum and the target
    def twoSumClosest(self, nums, target):
        # Write your code here
        nums.sort()
        i, j = 0, len(nums)  - 1

        diff = 0x7fffffff
        while i < j:
            if nums[i] + nums[j] < target:
                diff = min(diff, target - nums[i] - nums[j])
                i += 1
            else:
                diff = min(diff, nums[i] + nums[j] - target)
                j -= 1

        return diff

File: J9Ch/src/J_7_Two_Pointers/Required_10/python_cpp/_533_Two_Sum_Closest_to_target_Medium.cpp
/** 533 Two Sum - Closest to target
 * Medium*/


class Solution {
public:
    /**
     * @param nums an integer array
     * @param target an integer
     * @return the difference between the sum and the target
     */
    int twoSumClosest(vector<int>& nums, int target) {
        // Write your code here
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int j = n - 1;
        int diff = 0x7fffffff;
        for (int i = 0; i < n; ++i) {
            while (i < j && nums[i] + nums[j] > target) {
                if (nums[i] + nums[j] - target < diff)
                    diff = nums[i] + nums[j] - target;
                j --;
            }

            if (i >= j)
                break;

            if (target - nums[i] - nums[j] < diff)
                diff = target - nums[i] - nums[j];
        }

        return diff;
    }
};